ctmrg and variational optimisation

under-Peter opened the pull request.

under-Peter committed:
Implement differentiable trg using einsum and Zygote

GiggleLiu committed:
add more tests about result

GiggleLiu committed:
add the basic CTMRG system test

under-Peter committed:
Add manifest.toml to check whether tests pass

under-Peter committed:
Change repository URLs to maybe get travis to pass

under-Peter committed:
Update LinalgBackwards for julia 1.0 compatibility

under-Peter committed:
Working draft of ctmrg

under-Peter committed:
Test ctmrg by comparing with analytical results

To make the tests run fast, we keep bond-dimensions small and use 
generous tolerances in the comparisons. Due to random initialization of 
the corner and half-infinite row/column tensor we seed the RNG.

under-Peter committed:
Fix AD in ctmrg

GiggleLiu commented:
```
Test Summary: | Pass  Fail  Error  Total
ctmrg         |    1     5      1      7
```
Test fails on my laptop.

Is it because the check is too tight?

under-Peter commented:
> ```
> Test Summary: | Pass  Fail  Error  Total
> ctmrg         |    1     5      1      7
> ```
> 
> Test fails on my laptop.
> 
> Is it because the check is too tight?

They pass for me and on Travis. Not sure what the issue could be - if you run the tests, does it say there was an error or that the results weren't the expected ones?

under-Peter committed:
Change to BackwardsLinalg

under-Peter commented:
Tests pass now but documentation still fails - seems like because OMEinsum is not registered yet so we can just ignore that.

GiggleLiu commented:
I updated `Zygote` master and used the `dispatch-testing` branch.

Most issues resolved except this one

```julia
ctmrg: Test Failed at /home/leo/jcode/TensorNetworkAD.jl/test/ctmrg.jl:57
  Expression: magnetisationofβ(0, 2) ≈ magofβ(0)
   Evaluated: 6.223285321735374e-19 ≈ 0
Stacktrace:
 [1] top-level scope at /home/leo/jcode/TensorNetworkAD.jl/test/ctmrg.jl:57
 [2] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.1/Test/src/Test.jl:1083
 [3] top-level scope at /home/leo/jcode/TensorNetworkAD.jl/test/ctmrg.jl:56
```

under-Peter commented:
That's not surprising since `isapprox` shouldn't be used to compare with zero with default settings. It's strange though that it still works on my laptop and travis when it really shouldn't. I'll update it, thanks for noticing!

under-Peter committed:
Fix ctmrg test due to using isapprox with 0

GiggleLiu commented on src/ctmrg.jl:
```diff
@@ -0,0 +1,51 @@
+    t = initializet(a, χ)
+
+    # symmetrize
+    c += transpose(c)
```
Here, we use `einsum` instead of `transpose` and `permutedims`.

under-Peter committed:
Update package dependencies

under-Peter committed:
Integrate with OMEinsum

under-Peter committed:
Merge branch 'trg' into ctmrg

under-Peter committed:
Change to string-specification

under-Peter committed:
Remove custom einsum in favor of OMEinsum

under-Peter committed:
Move ising-stuff to its own folder and fix general things

under-Peter committed:
Rename variables and clean up in trg

under-Peter committed:
Add trg argument `tol` and use OMEinsum v0.1

under-Peter committed:
Fix AD

under-Peter committed:
Implement and test gradient based energy minimisation

under-Peter committed:
Add progressmeter for information during runtest

under-Peter committed:
Implement and test variationalipeps for tfising model

under-Peter committed:
Implement and test variationalpeps for heisenberg model

under-Peter committed:
Fix random seed for tests to pass

Probably could be more stable with options to `optimize` something to 
look into later.

under-Peter committed:
Improve progressmeter

under-Peter committed:
Add tests for rotational symmetry functions and gradient

under-Peter committed:
Add progressmeter and gradient-tests (currently broken)

under-Peter committed:
Seed random for gradient-check to pass

under-Peter committed:
Calculate lnZ per site directly to avoid exponential growth of value

under-Peter committed:
Provide keyword-arg to forward options to optimize

under-Peter committed:
Add option to change method of optimiseipeps

GiggleLiu commented on test/ctmrg.jl:
```diff
@@ -1,11 +1,31 @@
-    χ = 5
-    niter = 5
+    Random.seed!(1)
+    next!(pmobj)
```
do not use `next!(pmobj)` except the top level to make this subfile runable

under-Peter replied:
I comment out the tests I don't want to run in `runtest.jl` and run the usual test-suite with `]test`. This has the advantage that it uses a fresh environment for testing. Is that an option for you?

GiggleLiu replied:
Do you mean move this file to `examples/` rather than including them in tests?

Anyway, please make sure files included in `runtests.jl` being runnable on its own. Many people open these test files to see how it works.

under-Peter committed:
Change defaults, improve tests

under-Peter committed:
Introduce random and nonrandom initial condition for ctmrg

under-Peter committed:
Fix test

under-Peter committed:
Change ctmrg to use a fixedpoint interface

under-Peter committed:
Change to pairwise contractions

under-Peter committed:
Implement custom fixed-point-method backpropagation for ctmrg

under-Peter committed:
Remove tests of removed function

under-Peter committed:
Fix tests

a small perturbation needs to be added to some of the test-tensors for 
the diagonal hamiltonian to avoid lapack-errors.

under-Peter committed:
Minor fixes for potential complex inputs

under-Peter committed:
Get complex tensors to work

under-Peter committed:
fix missing testset

under-Peter committed:
Update dependencies

under-Peter committed:
Update to OMEinsum 0.2.0

mostly just changing the specifications of einsums to the string-literal 
form. In ctmrg, custom implementation of `norm` was used since otherwise 
Zygote introduced some nasty NaNs from time to time.

under-Peter committed:
Remove unused functions and add tests to improve coverage

under-Peter committed:
Insert Progressmeter again

this is to avoid travis timing out if there is no output for 10 minutes.

GiggleLiu commented on src/fixedpoint.jl:
```diff
@@ -27,35 +27,3 @@ function (st::StopFunction)(state)
 end
-
-function fixedpointbackward(next, (c,t,vals), (a, χ, d))
-    _, back = Zygote.forward(next,(c,t,vals),(a,χ,d))
```
why do you remove these code?

under-Peter replied:
It wasn't used and didn't work for the Complex case - it might belong on a feature branch but not in master

GiggleLiu commented on src/ctmrg.jl:
```diff
@@ -0,0 +1,78 @@
+    t += permutedims(conj(t), (3,2,1))
+
+    # normalize
+    c /= mynorm(c)
```
Is it possible to redefine the backward pass only, rather than writting a new function?

GiggleLiu replied:
Is this function differentiable on GPU?

under-Peter replied:
It doesn't work yet due to ambiguity errors - not sure if toehrwise it works.

and what do you mean with the first comment?

under-Peter replied:
oh, you mean for `norm`? I could just define the adjoint, but it's much easier to define the forward myself and let Zygote handle the rest - I doubt it's performance sensitive. Not sure about GPU since it does not run yet.

under-Peter replied:
I'll define just the adjoint - it's actually quite simple.

GiggleLiu commented on test/trg.jl:
```diff
@@ -1,15 +1,34 @@
-end
+    @testset "exampletensor" begin
+        β = rand()
+        @isdefined(pmobj) && next!(pmobj)
```
remove these lines, 
```julia
@isdefined(pmobj) && next!(pmobj)
```

many people learn a package from reading tests, making test cleaner is better for reading.

under-Peter committed:
Add docstrings

GiggleLiu commented on src/ctmrg.jl:
```diff
@@ -0,0 +1,117 @@
+using BackwardsLinalg, OMEinsum
```
Have already used `BackwardsLinalg` and `OMEinsum` in `TensorNteworkAD.jl`. Maybe move `using xxx` statements to `TensorNteworkAD.jl`?

GiggleLiu commented on src/ctmrg.jl:
```diff
@@ -0,0 +1,117 @@
+differences between `c`s singular values between two steps is
+below `tol` the algorithm is assumed to be converged.
+"""
+function ctmrg(a, χ, tol, maxit::Integer, randinit = false)
```
Better specify types for a,\chi and  tol

GiggleLiu replied:
One of my favorite design of such iterative algorithms is implementing `Base.iterate`, like https://github.com/QuantumBFS/Pathintegral-QMC.jl/blob/9970e9392b585e546131f640b3ae283272916a7a/src/PathIntegral.jl#L114 , then we do not need `StopFunction` or `callback`s. No need to change here.

GiggleLiu replied:
Please allow the initial tensor `T0` as input rather than specifying `randinit`.

A using case: one can backup the einvironment `tensor` to disk every 100 step. In case the program breaks by accident, he can use this tensor as input to continue training.

Another using case: we can use a converged tensor at hamiltonian parameter `h+dh` as the initial tensor for the training at parameter `h`.

under-Peter replied:
I also like the base iterate design (used it in my other cornertransfer-package) but for simplicity I used the more naive approach here.

Also adding the possibility to provide initial c/t

GiggleLiu commented on src/ctmrg.jl:
```diff
@@ -0,0 +1,117 @@
+
+    # renormalize
+    cpmat = reshape(cp, χ*d, χ*d)
+    cpmat += adjoint(cpmat)
```
just our of curiosity, why not using `cpmat += cpmat'`

under-Peter replied:
no reason - I'll change it

GiggleLiu commented on src/ctmrg.jl:
```diff
@@ -0,0 +1,117 @@
+
+    # symmetrize
+    c += adjoint(c)
+    t += permutedims(conj(t), (3,2,1))
```
use `ein"ijk->kji"(conj(t))`, as a pressure test for `OMEinsum`.

GiggleLiu commented on src/exampletensors.jl:
```diff
@@ -0,0 +1,63 @@
+
+"""
+    tensorfromclassical(h)
+given a classical 2-body hamiltonian `h`, return the corresponding tensor
```
Show the type and definition of `h`.

GiggleLiu commented on src/exampletensors.jl:
```diff
@@ -0,0 +1,63 @@
+return the isingtensor at inverse temperature `β` for a two-dimensional
+grid tensor-network.
+"""
+function isingtensor(β)
```
We need some abstraction here, the following programming style is more flexible
```
mag(::Ising)
initial_tensor(::Ising)
```

`Ising` is a model type that you can dispatch on. You can create new model easily by dispatching these methods.

under-Peter replied:
since I only provide the isingmodel, I'm not sure that the abstraction is necessary.

If I did it anyway, what would you suggest more exactly? a `mag(::Ising,beta)`? or `mag(::Ising)(beta)` or something else?

GiggleLiu replied:
`mag(model, beta)` is good. Model can be a `Ising` instance.

GiggleLiu commented on src/fixedpoint.jl:
```diff
@@ -0,0 +1,40 @@
+    end
+end
+
+mutable struct StopFunction{T,S}
```
Don't need a structure for this purpose here. A lambda function would be enough?

under-Peter replied:
the structure is necessary to save the old svd-values for comparison. A closure could work too but the struct is easier I think.

GiggleLiu replied:
ok, I see. :+1:

GiggleLiu commented on src/variationalipeps.jl:
```diff
@@ -0,0 +1,110 @@
+return a normalized `x` whith left-right,
+up-down, diagonal and rotational symmetry.
+"""
+function symmetrize(x::AbstractArray{<:Any,5})
```
use a more specific name, `symmetrize` is too general.

under-Peter committed:
ctmrg now accepts initial corner and row/column tensors as arguments

under-Peter committed:
Remove testprogressmeter

GiggleLiu commented on src/variationalipeps.jl:
```diff
@@ -0,0 +1,110 @@
+return the transverse field ising hamiltonian with transverse magnetic
+field `hx` as a two-site operator.
+"""
+function tfisinghamiltonian(hx::Float64 = 1.0)
```
`hamiltonian(::TFIIsing)` and `hamiltonian(::Heisenberg)`looks more extensible.

GiggleLiu commented on src/variationalipeps.jl:
```diff
@@ -0,0 +1,110 @@
+providing `optimmethod`. Other options to optim can be passed with `optimargs`.
+The energy is calculated using ctmrg with parameters `χ`, `tol` and `maxit`.
+"""
+function optimiseipeps(t, h, χ, tol, maxit;
```
Maybe using `optimizer` as input paramter is better. Current interface does not allow people to tune the optimizer.

under-Peter replied:
Given how simple `optimiseipeps` is, if a different optimizer is used one should probably just rewrite it. Like this it's easy to have intermediate results printed (via optimargs) or chose a different algorithm from Optim but there are also sane defaults. At least for my experimenting, it worked well.

GiggleLiu commented on test/runtests.jl:
```diff
@@ -1,7 +1,14 @@
+
+Random.seed!(4)
 
 @testset "TensorNetworkAD.jl" begin
```
None of these tests are unit tests. :disappointed_relieved:  . Without unit tests, programs can easily get incorrect results when generaling it to other instances.

under-Peter committed:
generalize model-tensor, magnetisation-tensor and hamiltonian interface

under-Peter committed:
add trg_svd unittest

under-Peter committed:
replace norm-adjoint for Zygote and add tests

under-Peter committed:
Add more tests

GiggleLiu commented:
Except the `optimiseipeps` interface issue, the rest are good. Will merge it first, we can still polish it after this summer.

Well done, cheers!
